src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractSyncNamespaceService.State.cs

```diff
@@ -0,0 +1,297 @@
+    {
+        internal sealed class State
+        {                                                                        
+            private static readonly SymbolDisplayFormat s_qualifiedNameOnlyFormat = 
+                new SymbolDisplayFormat(globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+                                        typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces);
+
+            public Solution Solution { get; }
+
+            public DocumentId OriginalDocumentId { get; }
+
+            public ImmutableArray<DocumentId> DocumentIds { get;  }
+
+            /// <summary>
+            /// This is the default namespace defined in the project file.
+            /// </summary>
+            public string DefaultNamespace { get; }
+
+            /// <summary>
+            /// This is the name of the namespace declaration that trigger the refactoring.
+            /// </summary>
+            public string DeclaredNamespace { get; }
+                                                                                                       
+            /// <summary>
+            /// This is the new name we want to change the namespace to.
+            /// Empty string means global namespace, whereas null means renaming action is not available.
+            /// </summary>
+            public string TargetNamespace { get; }
+
+            /// <summary>
+            /// This is the part of the declared namespace that is contained in default namespace.
+            /// We will use this to construct target folder to move the file to.
+            /// For example, if default namespace is `A` and declared namespace is `A.B.C`, 
+            /// this would be `B.C`.
+            /// </summary>
+            public string RelativeDeclaredNamespace { get; }
+
+            private State(
+                Solution solution,
+                DocumentId originalDocumentId,
+                ImmutableArray<DocumentId> documentIds,
+                string rootNamespce,
+                string targetNamespace,
+                string declaredNamespace,                           
+                string relativeDeclaredNamespace)
+            {
+                Solution = solution;
+                OriginalDocumentId = originalDocumentId;
+                DocumentIds = documentIds;
+                DefaultNamespace = rootNamespce;
+                TargetNamespace = targetNamespace;
+                DeclaredNamespace = declaredNamespace;
+                RelativeDeclaredNamespace = relativeDeclaredNamespace;
+            }
+
+            /// <summary>
+            /// This refactoring only supports non-linked document and linked document in the form of
+            /// documents in multi-targeting project. Also for simplicity, we also don't support document
+            /// what has different file path and logical path in project (i.e. [ProjectRoot] + `Document.Folders`). 
```

Comment:
i don't understand this case.

Reply:
i still don't understand this case.

Reply:
This covers a few cases:
1. When a file is a linked file via `<link>` in project file, and is the only instance of that file in the solution, then `GetLinkedDocumentIds` will return nothing. We want to ignore such files. Of course, it's possible someone decided to has such a linked file with exact same logical path and actual path on disk, then this check will let it through and trigger the refactoring. But such case should be very rare (there's a todo inside the method for this)
2. A project can have a file not rooted in project root. For example, `c:\projectA\projectA.csproj` might contain a file `c:\foo.cs`. We'd like to filter those out as well.